### ISSUE TITLE
ExtractMethodCodeRefactoringProvider encountered an error and has been disabled.

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#31249]
Reported in 15.9 

System.InvalidOperationException : Matrix does not contain Key 'AB'.
   at Roslyn.Utilities.Contract.ThrowIfFalse(Boolean condition,String message)
   at Microsoft.CodeAnalysis.ExtractMethod.ExtractMethodMatrix.GetVariableStyle(Boolean captured,Boolean dataFlowIn,Boolean dataFlowOut,Boolean alwaysAssigned,Boolean variableDeclared,Boolean readInside,Boolean writtenInside,Boolean readOutside,Boolean writtenOutside,Boolean unsafeAddressTaken)
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GetVariableStyle(Dictionary`2 symbolMap,ISymbol symbol,SemanticModel model,ITypeSymbol type,Boolean captured,Boolean dataFlowIn,Boolean dataFlowOut,Boolean alwaysAssigned,Boolean variableDeclared,Boolean readInside,Boolean writtenInside,Boolean readOutside,Boolean writtenOutside,Boolean unsafeAddressTaken)
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GenerateVariableInfoMap(SemanticModel model,DataFlowAnalysis dataFlowAnalysisData,Dictionary`2 symbolMap)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.AnalyzeAsync(&lt;Unknown Parameters&gt;)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.ExtractMethodAsync(&lt;Unknown Parameters&gt;)
   at async Microsoft.CodeAnalysis.ExtractMethod.AbstractExtractMethodService`3.ExtractMethodAsync[TValidator,TExtractor,TResult](&lt;Unknown Parameters&gt;)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.GetCodeActionAsync(&lt;Unknown Parameters&gt;)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.ComputeRefactoringsAsync(&lt;Unknown Parameters&gt;)
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.GetRefactoringFromProviderAsync(&lt;Unknown Parameters&gt;)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/385359/extractmethodcoderefactoringprovider-encountered-a-16.html
VSTS ticketId: 731751_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_
(no solutions)

**Comment by jinujoseph** [comment]
Related to https://github.com/dotnet/roslyn/issues/23352

**Comment by heejaechang** [comment]
@jinujoseph how can I follow up with the customer? I want to get repro steps from the users.

**Comment by jinujoseph** [comment]
you can leave the comment here or  turn the DC link to 'need more info' with your question.

**Comment by heejaechang** [comment]
@agocke what code can cause this states?

dataFlowIn: true, 
dataFlowOut: false, 
alwaysAssigned: true, 
variableDeclared: false, 
readInside: true, 
writtenInside: false, 
readOutside: true, 
writtenOutside: true

... 

basically, question is what always assigned is "true", but written inside is "false"

**Comment by heejaechang** [comment]
@gafter can you help me to figure out what kind of code can cause data flow analysis to return the above state?

**Comment by CyrusNajmabadi** [comment]
@tamlin-mike I've fixed up the behavior of ExtractMethod here: https://github.com/dotnet/roslyn/pull/30554

We now will no longer crash in these cases, and we'll successfully be able to extract.  You should see the fix in 16.0.  Cheers!

**Comment by nnpcYvIVl** [comment]
@heejaechang Here's a repro if you still need one.

Double click on the call to InnerMethod() to cause the crash.
```csharp
using System;

namespace Roslyn_GetVariableStyle_InvalidOperationException
{
    public static class Class
    {
        public static void Method(string value)
        {
            IntPtr intptr = IntPtr.Zero;

            void InnerMethod()
            {
                try
                {
                    intptr = AnotherMethod(value);
                }
                finally
                {
                    if (intptr != IntPtr.Zero) { }
                }
            }

            InnerMethod();
        }
    }
}
```
![roslyn getvariablestyle invalidoperationexception](https://user-images.githubusercontent.com/8421414/51625881-f0fd3600-1f03-11e9-9482-394b2dec61fa.png)
